include/fiction/algorithms/simulation/sidb/ground_state_space.hpp

```diff
@@ -0,0 +1,802 @@
+     * top cluster as the number of combinations of \f$N\f$ stars and \f$b - 1\f$ bars. Hence this is computed with the
+     * following combinatorial formula: \f$\binom{N + b - 1}{b - 1}\f$.
+     */
+    const uint64_t maximum_top_level_multisets{};
+};
+
+namespace detail
+{
+
+template <typename Lyt>
+class ground_state_space_impl
+{
+  public:
+    explicit ground_state_space_impl(const Lyt& lyt, const uint64_t max_cluster_size_for_witness_partitioning,
+                                     const sidb_simulation_parameters& phys_params) noexcept :
+            base{phys_params.base},
+            top_cluster{to_sidb_cluster(sidb_cluster_hierarchy(lyt))},
+            clustering{get_initial_clustering(top_cluster, get_local_potential_bounds(lyt, phys_params))},
+            witness_partitioning_max_cluster_size{max_cluster_size_for_witness_partitioning},
+            mu_bounds_with_error{physical_constants::POP_STABILITY_ERR - phys_params.mu_minus,
+                                 -physical_constants::POP_STABILITY_ERR - phys_params.mu_minus,
+                                 physical_constants::POP_STABILITY_ERR - phys_params.mu_plus(),
+                                 -physical_constants::POP_STABILITY_ERR - phys_params.mu_plus()}
+    {}
+
+    ground_state_space_result run() noexcept
+    {
+        mockturtle::stopwatch<>::duration time_counter{};
+        {
+            const mockturtle::stopwatch stop{time_counter};
+
+            while (!terminate)
+            {
+                while (!update_charge_spaces())
+                    ;
+
+                move_up_hierarchy();
+            }
+        }
+
+        const uint64_t max_multisets = maximum_top_level_multisets(top_cluster->size());
+
+        return ground_state_space_result{top_cluster, time_counter, max_multisets - top_cluster->charge_space.size(),
+                                         max_multisets};
+    }
+
+  private:
+    constexpr inline bool fail_onto_negative_charge(const double pot_bound) const noexcept
+    {
+        // V > e - mu-
+        return pot_bound > mu_bounds_with_error[0];
+    }
+
+    constexpr inline bool fail_onto_positive_charge(const double pot_bound) const noexcept
+    {
+        // V < -e - mu+
+        return pot_bound < mu_bounds_with_error[3];
+    }
+
+    constexpr inline bool ub_fail_onto_neutral_charge(const double pot_bound) const noexcept
+    {
+        // V < -e - mu-
+        return pot_bound < mu_bounds_with_error[1];
+    }
+
+    constexpr inline bool lb_fail_onto_neutral_charge(const double pot_bound) const noexcept
+    {
+        // V > e - mu+
+        return pot_bound > mu_bounds_with_error[2];
+    }
+
+    static std::pair<charge_distribution_surface<Lyt>, charge_distribution_surface<Lyt>>
+    get_local_potential_bounds(const Lyt& lyt, const sidb_simulation_parameters& phys_params) noexcept
+    {
+        charge_distribution_surface<Lyt> cl_min{lyt};
```

Comment:
Typically, we use ``cds`` instead of ``cl``.

Reply:
done